_posts/2018-05-02-load-balanced-smart-proxies-with-haproxy.md

```diff
@@ -0,0 +1,138 @@
+The HAProxy setup itself is super simple, just load-balance a bunch of ports in TCP mode round-robin across all the Smart Proxies. TCP mode is required because the load balancer will not be doing TLS termination. There are just two exceptions to that: port 443 should use `source` not `roundrobin` to not confuse Yum, and the additional port for the PuppetCA (8141) is forwarded only to the first smart proxy.
+
+| Service    | Port | Mode | Balance mode | Destination |
+| ---------- | ---- | ---- | ------------ | ----------- |
+| HTTP       | 80   | TCP  | `roundrobin` | port 80 on all proxies |
+| Anaconda   | 8000 | TCP  | `roundrobin` | port 8000 on all proxies|
+| HTTPS      | 443  | TCP  | `source`     | port 443 on all proxies |
+| RHSM       | 8443 | TCP  | `roundrobin` | port 8443 on all proxies |
+| AMQP       | 5647 | TCP  | `roundrobin` | port 5647 on all proxies |
+| Puppet     | 8140 | TCP  | `roundrobin` | port 8140 on all proxies |
+| PuppetCA   | 8141 | TCP  | `roundrobin` | port 8140 on the first proxy |
+| SmartProxy | 9090 | TCP  | `roundrobin` | port 9090 on all proxies |
+| Docker     | 5000 | TCP  | `roundrobin` | port 5000 on all proxies |
+
+[You can find an example `haproxy.cfg` in the `forklift` repository.](https://github.com/theforeman/forklift/blob/master/roles/haproxy/templates/haproxy.cfg.j2)
+
+## Configuring Clients
+
+As far as clients are concerned, they should only know about `smartproxy.example.test`, and should never need to connect directly to a smartproxy. That has to be configured for Puppet, goferd, and for RHSM/Yum.
+
+### Using katello-client-bootstrap
+
+[katello-client-bootstrap](https://github.com/Katello/katello-client-bootstrap) is designed to configure a machine to be used with an exiting Katello environment. It also can configure Puppet properly for us, and supports [setting a separate Puppet CA and Puppet CA port in current Git](https://github.com/Katello/katello-client-bootstrap/pull/250).
+
+To attach clients to the load-balanced proxy, we just have to pass `--server smartproxy.example.test` and `--puppet-ca-port 8141`, `bootstrap.py` will take care of the rest.
+
+### Manual registration
+
+Instead of using `bootstrap.py`, the client can also be registered manually. However, we recommend using `bootstrap.py` since it handles these extra manual steps for you.
+
+After installing `katello-ca-consumer` RPM from `http://smartproxy.example.test/pub/katello-ca-consumer-latest.noarch.rpm` `/etc/rhsm/rhsm.conf` and `/etc/yum.repos.d/redhat.repo` will contain the hostname of whichever proxy served the initial request and not `smartproxy.example.net` as we'd like it. Thus, when registering the system with `subscription-manager`, we have to pass `--serverurl=https://smartproxy.example.test:8443/rhsm --baseurl=https://smartproxy.example.test/pulp/repos` to update those two config files.
+
+Also Puppet needs to be configured to talk to the split-out PuppetCA, so the `puppet.conf` needs the following lines:
```

Comment:
Wait ... to refer to the "split-out PuppetCA" is kind of irritating me. Is it relevant, that its split out at this point? 

I guess not, because the haproxy configuration handles forwarding to the right server, so I'd write that PuppetCA has to be configured to use the shared hostname, but not explicitly refer that detail of the balancer setup.

Besides that just a wording suggestion: s/Also/Additionally/ ... but "Also" is probably okay.

Reply:
Well, it needs to talk to the dedicated port, as that is what is forwarded to proxy01 only. I rephrased that sentence, does it make more sense now?

Reply:
Fair enough, I somehow missed (to keep in mind) the fact that PuppetCA is running on a different port. I would still suggest a rephrasing. Maybe something like "Since our PuppetCA is reachable on a dedicated port, we additionally need to adjust the puppet.conf to use that port and the shared hostname for certificate requests."

Reply:
Aye, updated to include something very similar to your line :)